File: backend_py/my_agent/DataFormatter.py
import json
from langchain_core.prompts import ChatPromptTemplate
from my_agent.LLMManager import LLMManager
from my_agent.graph_instructions import graph_instructions


class DataFormatter:
    def __init__(self):
        self.llm_manager = LLMManager()

    
    def format_data_for_visualization(self, state: dict) -> dict:
        """Format the data for the chosen visualization type."""
        visualization = state['visualization']
        results = state['results']
        question = state['question']
        sql_query = state['sql_query']

        if visualization == "none":
            return {"formatted_data_for_visualization": None}
        
        if visualization == "scatter":
            try:
                return self._format_scatter_data(results)
            except Exception as e:
                return self._format_other_visualizations(visualization, question, sql_query, results)
        
        if visualization == "bar" or visualization == "horizontal_bar":
            try:
                return self._format_bar_data(results, question)
            except Exception as e:
                return self._format_other_visualizations(visualization, question, sql_query, results)
        
        if visualization == "line":
            try:
                return self._format_line_data(results, question)
            except Exception as e:
                return self._format_other_visualizations(visualization, question, sql_query, results)
        
        return self._format_other_visualizations(visualization, question, sql_query, results)
    
    def _format_line_data(self, results, question):
        if isinstance(results, str):
            results = eval(results)

        if len(results[0]) == 2:

            x_values = [str(row[0]) for row in results]
            y_values = [float(row[1]) for row in results]

            # Use LLM to get a relevant label
            prompt = ChatPromptTemplate.from_messages([
                ("system", "You are a data labeling expert. Given a question and some data, provide a concise and relevant label for the data series."),
                ("human", "Question: {question}\n Data (first few rows): {data}\n\nProvide a concise label for this y axis. For example, if the data is the sales figures over time, the label could be 'Sales'. If the data is the population growth, the label could be 'Population'. If the data is the revenue trend, the label could be 'Revenue'."),
            ])
            label = self.llm_manager.invoke(prompt, question=question, data=str(results[:2]))

            formatted_data = {
                "xValues": x_values,
                "yValues": [
                    {
                        "data": y_values,
                        "label": label.strip()
                    }
                ]
            }
        elif len(results[0]) == 3:

            # Group data by label
            data_by_label = {}
            x_values = []

            # Get a list of unique labels
            labels = list(set(item2 for item1, item2, item3 in results 
                              if isinstance(item2, str) and not item2.replace(".", "").isdigit() and "/" not in item2))
            
            # If labels are not in the second position, check the first position
            if not labels:
                labels = list(set(item1 for item1, item2, item3 in results 
                                  if isinstance(item1, str) and not item1.replace(".", "").isdigit() and "/" not in item1))

            for item1, item2, item3 in results:
                # Determine which item is the label (string not convertible to float and not containing "/")
                if isinstance(item1, str) and not item1.replace(".", "").isdigit() and "/" not in item1:
                    label, x, y = item1, item2, item3
                else:
                    x, label, y = item1, item2, item3
                    

                if str(x) not in x_values:
                    x_values.append(str(x))
                if label not in data_by_label:
                    data_by_label[label] = []
                data_by_label[label].append(float(y))
                print(labels)
                for other_label in labels:
                    if other_label != label:
                        if other_label not in data_by_label:
                            data_by_label[other_label] = []
                        data_by_label[other_label].append(None)

            # Create yValues array
            y_values = [
                {
                    "data": data,
                    "label": label
                }
                for label, data in data_by_label.items()
            ]

            formatted_data = {
                "xValues": x_values,
                "yValues": y_values,
                "yAxisLabel": ""
            }

            # Use LLM to get a relevant label for the y-axis
            prompt = ChatPromptTemplate.from_messages([
                ("system", "You are a data labeling expert. Given a question and some data, provide a concise and relevant label for the y-axis."),
                ("human", "Question: {question}\n Data (first few rows): {data}\n\nProvide a concise label for the y-axis. For example, if the data represents sales figures over time for different categories, the label could be 'Sales'. If it's about population growth for different groups, it could be 'Population'."),
            ])
            y_axis_label = self.llm_manager.invoke(prompt, question=question, data=str(results[:2]))

            # Add the y-axis label to the formatted data
            formatted_data["yAxisLabel"] = y_axis_label.strip()

        return {"formatted_data_for_visualization": formatted_data}

    def _format_scatter_data(self, results):
        if isinstance(results, str):
            results = eval(results)

        formatted_data = {"series": []}
        
        if len(results[0]) == 2:
            formatted_data["series"].append({
                "data": [
                    {"x": float(x), "y": float(y), "id": i+1}
                    for i, (x, y) in enumerate(results)
                ],
                "label": "Data Points"
            })
        elif len(results[0]) == 3:
            entities = {}
            for item1, item2, item3 in results:
                # Determine which item is the label (string not convertible to float and not containing "/")
                if isinstance(item1, str) and not item1.replace(".", "").isdigit() and "/" not in item1:
                    label, x, y = item1, item2, item3
                else:
                    x, label, y = item1, item2, item3
                if label not in entities:
                    entities[label] = []
                entities[label].append({"x": float(x), "y": float(y), "id": len(entities[label])+1})
            
            for label, data in entities.items():
                formatted_data["series"].append({
                    "data": data,
                    "label": label
                })
        else:
            raise ValueError("Unexpected data format in results")                

        return {"formatted_data_for_visualization": formatted_data}


    def _format_bar_data(self, results, question):
        if isinstance(results, str):
            results = eval(results)

        if len(results[0]) == 2:
            # Simple bar chart with one series
            labels = [str(row[0]) for row in results]
            data = [float(row[1]) for row in results]
            
            # Use LLM to get a relevant label
            prompt = ChatPromptTemplate.from_messages([
                ("system", "You are a data labeling expert. Given a question and some data, provide a concise and relevant label for the data series."),
                ("human", "Question: {question}\nData (first few rows): {data}\n\nProvide a concise label for this y axis. For example, if the data is the sales figures for products, the label could be 'Sales'. If the data is the population of cities, the label could be 'Population'. If the data is the revenue by region, the label could be 'Revenue'."),
            ])
            label = self.llm_manager.invoke(prompt, question=question, data=str(results[:2]))
            
            values = [{"data": data, "label": label}]
        elif len(results[0]) == 3:
            # Grouped bar chart with multiple series
            categories = set(row[1] for row in results)
            labels = list(categories)
            entities = set(row[0] for row in results)
            values = []
            for entity in entities:
                entity_data = [float(row[2]) for row in results if row[0] == entity]
                values.append({"data": entity_data, "label": str(entity)})
        else:
            raise ValueError("Unexpected data format in results")

        formatted_data = {
            "labels": labels,
            "values": values
        }

        return {"formatted_data_for_visualization": formatted_data}

    def _format_other_visualizations(self, visualization, question, sql_query, results):
        instructions = graph_instructions[visualization]
        prompt = ChatPromptTemplate.from_messages([
            ("system", "You are a Data expert who formats data according to the required needs. You are given the question asked by the user, it's sql query, the result of the query and the format you need to format it in."),
            ("human", 'For the given question: {question}\n\nSQL query: {sql_query}\n\Result: {results}\n\nUse the following example to structure the data: {instructions}. Just give the json string. Do not format it'),
        ])
        response = self.llm_manager.invoke(prompt, question=question, sql_query=sql_query, results=results, instructions=instructions)
            
        try:
            formatted_data_for_visualization = json.loads(response)
            return {"formatted_data_for_visualization": formatted_data_for_visualization}
        except json.JSONDecodeError:
            return {"error": "Failed to format data for visualization", "raw_response": response}

File: backend_js/my_agent/DataFormatter.ts
import { ChatPromptTemplate } from "@langchain/core/prompts";
import LLMManager from "./LLMManager";
import { graph_instructions } from "./graph_instructions";

class DataFormatter {
  private llm_manager: LLMManager;

  constructor() {
    this.llm_manager = new LLMManager();

    this.format_data_for_visualization =
      this.format_data_for_visualization.bind(this);
    this._format_scatter_data = this._format_scatter_data.bind(this);
    this._format_bar_data = this._format_bar_data.bind(this);
    this._format_line_data = this._format_line_data.bind(this);
    this._format_other_visualizations =
      this._format_other_visualizations.bind(this);
  }

  async format_data_for_visualization(
    state: Record<string, any>
  ): Promise<Record<string, any>> {
    const visualization = state["visualization"];
    const results = state["results"];
    const question = state["question"];
    const sql_query = state["sql_query"];

    if (visualization === "none") {
      return { formatted_data_for_visualization: null };
    }

    if (visualization === "scatter") {
      try {
        return await this._format_scatter_data(results);
      } catch (e) {
        return await this._format_other_visualizations(
          visualization,
          question,
          sql_query,
          results
        );
      }
    }

    if (visualization === "bar" || visualization === "horizontal_bar") {
      try {
        return await this._format_bar_data(results, question);
      } catch (e) {
        return await this._format_other_visualizations(
          visualization,
          question,
          sql_query,
          results
        );
      }
    }

    if (visualization === "line") {
      try {
        return await this._format_line_data(results, question);
      } catch (e) {
        return await this._format_other_visualizations(
          visualization,
          question,
          sql_query,
          results
        );
      }
    }

    return await this._format_other_visualizations(
      visualization,
      question,
      sql_query,
      results
    );
  }

  private async _format_line_data(
    results: any,
    question: string
  ): Promise<Record<string, any>> {
    if (typeof results === "string") {
      results = JSON.parse(results);
    }

    if (results[0].length === 2) {
      const x_values = results.map((row: any[]) => String(row[0]));
      const y_values = results.map((row: any[]) => parseFloat(row[1]));

      const prompt = ChatPromptTemplate.fromMessages([
        [
          "system",
          "You are a data labeling expert. Given a question and some data, provide a concise and relevant label for the data series.",
        ],
        [
          "human",
          "Question: {question}\n Data (first few rows): {data}\n\nProvide a concise label for this y axis. For example, if the data is the sales figures over time, the label could be 'Sales'. If the data is the population growth, the label could be 'Population'. If the data is the revenue trend, the label could be 'Revenue'.",
        ],
      ]);
      const label = await this.llm_manager.invoke(prompt, {
        question,
        data: JSON.stringify(results.slice(0, 2)),
      });

      const formatted_data = {
        xValues: x_values,
        yValues: [
          {
            data: y_values,
            label: label.trim(),
          },
        ],
      };
      return { formatted_data_for_visualization: formatted_data };
    } else if (results[0].length === 3) {
      const data_by_label: Record<string, any[]> = {};
      const x_values: string[] = [];

      // Get a list of unique labels
      const labels = [
        ...new Set(
          results
            .filter(
              ([item1, item2]: [any, any]) =>
                (typeof item2 === "string" &&
                  isNaN(parseFloat(item2)) &&
                  !item2.includes("/")) ||
                (typeof item1 === "string" &&
                  isNaN(parseFloat(item1)) &&
                  !item1.includes("/"))
            )
            .map(([item1, item2]: [any, any]) =>
              typeof item2 === "string" &&
              isNaN(parseFloat(item2)) &&
              !item2.includes("/")
                ? item2
                : item1
            )
        ),
      ];

      for (const [item1, item2, item3] of results) {
        let label, x, y;
        // Determine which item is the label (string not convertible to float and not containing "/")
        if (
          typeof item1 === "string" &&
          isNaN(parseFloat(item1)) &&
          !item1.includes("/")
        ) {
          [label, x, y] = [item1, item2, item3];
        } else {
          [x, label, y] = [item1, item2, item3];
        }

        if (!x_values.includes(String(x))) {
          x_values.push(String(x));
        }
        if (!data_by_label[label]) {
          data_by_label[label] = [];
        }
        data_by_label[label].push(parseFloat(y));

        for (const other_label of labels) {
          if (other_label !== label) {
            if (!data_by_label[other_label as string]) {
              data_by_label[other_label as string] = [];
            }
            data_by_label[other_label as string].push(null);
          }
        }
      }

      // Create yValues array
      const y_values = Object.entries(data_by_label).map(([label, data]) => ({
        data,
        label,
      }));

      const formatted_data = {
        xValues: x_values,
        yValues: y_values,
        yAxisLabel: "",
      };

      // Use LLM to get a relevant label for the y-axis
      const prompt = ChatPromptTemplate.fromMessages([
        [
          "system",
          "You are a data labeling expert. Given a question and some data, provide a concise and relevant label for the y-axis.",
        ],
        [
          "human",
          "Question: {question}\n Data (first few rows): {data}\n\nProvide a concise label for the y-axis. For example, if the data represents sales figures over time for different categories, the label could be 'Sales'. If it's about population growth for different groups, it could be 'Population'.",
        ],
      ]);
      const y_axis_label = await this.llm_manager.invoke(prompt, {
        question: question,
        data: JSON.stringify(results.slice(0, 2)),
      });

      formatted_data["yAxisLabel"] = y_axis_label.trim();

      return { formatted_data_for_visualization: formatted_data };
    }

    throw new Error("Unexpected data format in results");
  }

  private async _format_scatter_data(
    results: any
  ): Promise<Record<string, any>> {
    if (typeof results === "string") {
      results = JSON.parse(results);
    }

    const formatted_data: { series: any[] } = { series: [] };

    if (results[0].length === 2) {
      formatted_data.series.push({
        data: results.map(([x, y]: [number, number], i: number) => ({
          x: x,
          y: y,
          id: i + 1,
        })),
        label: "Data Points",
      });
    } else if (results[0].length === 3) {
      const entities: Record<string, any[]> = {};
      for (const [item1, item2, item3] of results) {
        let label: string, x: number, y: number;
        if (
          typeof item1 === "string" &&
          isNaN(parseFloat(item1)) &&
          !item1.includes("/")
        ) {
          [label, x, y] = [item1, item2, item3];
        } else {
          [x, label, y] = [item1, item2, item3];
        }
        if (!entities[label]) {
          entities[label] = [];
        }
        entities[label].push({
          x: x,
          y: y,
          id: entities[label].length + 1,
        });
      }

      for (const [label, data] of Object.entries(entities)) {
        formatted_data.series.push({
          data,
          label,
        });
      }
    } else {
      throw new Error("Unexpected data format in results");
    }

    return { formatted_data_for_visualization: formatted_data };
  }

  private async _format_bar_data(
    results: any,
    question: string
  ): Promise<Record<string, any>> {
    if (typeof results === "string") {
      results = JSON.parse(results);
    }

    let formatted_data: Record<string, any>;

    if (results[0].length === 2) {
      const labels = results.map((row: any[]) => String(row[0]));
      const data = results.map((row: any[]) => parseFloat(row[1]));

      const prompt = ChatPromptTemplate.fromMessages([
        [
          "system",
          "You are a data labeling expert. Given a question and some data, provide a concise and relevant label for the data series.",
        ],
        [
          "human",
          "Question: {question}\nData (first few rows): {data}\n\nProvide a concise label for this y axis. For example, if the data is the sales figures for products, the label could be 'Sales'. If the data is the population of cities, the label could be 'Population'. If the data is the revenue by region, the label could be 'Revenue'.",
        ],
      ]);
      const label = await this.llm_manager.invoke(prompt, {
        question,
        data: JSON.stringify(results.slice(0, 2)),
      });

      formatted_data = {
        labels,
        values: [{ data, label }],
      };
    } else if (results[0].length === 3) {
      const categories = new Set(results.map((row: any[]) => row[1]));
      const labels = Array.from(categories);
      const entities = new Set(results.map((row: any[]) => row[0]));
      const values = Array.from(entities).map((entity) => ({
        data: results
          .filter((row: any[]) => row[0] === entity)
          .map((row: any[]) => parseFloat(row[2])),
        label: String(entity),
      }));

      formatted_data = { labels, values };
    } else {
      throw new Error("Unexpected data format in results");
    }

    return { formatted_data_for_visualization: formatted_data };
  }

  private async _format_other_visualizations(
    visualization: string,
    question: string,
    sql_query: string,
    results: any
  ): Promise<Record<string, any>> {
    const instructions =
      graph_instructions[visualization as keyof typeof graph_instructions];
    const prompt = ChatPromptTemplate.fromMessages([
      [
        "system",
        "You are a Data expert who formats data according to the required needs. You are given the question asked by the user, it's sql query, the result of the query and the format you need to format it in.",
      ],
      [
        "human",
        "For the given question: {question}\n\nSQL query: {sql_query}\nResult: {results}\n\nUse the following example to structure the data: {instructions}. Just give the json string. Do not format it",
      ],
    ]);
    const response = await this.llm_manager.invoke(prompt, {
      question,
      sql_query,
      results: JSON.stringify(results),
      instructions,
    });

    try {
      const formatted_data_for_visualization = JSON.parse(response);
      return { formatted_data_for_visualization };
    } catch (error) {
      return {
        error: "Failed to format data for visualization",
        raw_response: response,
      };
    }
  }
}

export default DataFormatter;
